### PR TITLE
ENH: implement pandas_timedelta_to_timedeltastruct for other resos

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the Github issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

cc @WillAyd IIRC you implemented the original here.  The new cases this implements are mostly copy/paste, so it wouldn't surprise me if there is a more elegant alternative.  Thoughts?

### CODE REVIEW

File: pandas/_libs/tslibs/src/datetime/np_datetime.c [hunk 1]
```diff
@@ -682,19 +682,23 @@ void pandas_timedelta_to_timedeltastruct(npy_timedelta td,
     npy_int64 sfrac;
     npy_int64 ifrac;
     int sign;
-    npy_int64 DAY_NS = 86400000000000LL;
+    npy_int64 PER_DAY;
```
Comment (WillAyd) [inline pandas/_libs/tslibs/src/datetime/np_datetime.c:685]: Just as a matter of convention should probably be lowercase. Usually you uppercase macros and constants, so on first read looking at these in the body of the functions a reader may mistakenly think that (DAY_NS was incorrect)

**Review by WillAyd — COMMENTED** [review]
Nice! Glad to see you venture down the C path. This is a great start.

On first pass seems reasonable - your plan is to ultimately generate these for all precisions right? If so you might want to opt for a macro'ed approach similar to what's in khash.h rather than copy / pasting these.

Here's an example of what that might look like

https://github.com/pandas-dev/pandas/blob/5414a2d59d53c4340e19db68beaea5856c1501a7/pandas/_libs/src/klib/khash.h#L708

File: pandas/_libs/tslibs/src/datetime/np_datetime.c (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/src/datetime/np_datetime.c:685]: good to know, will update

**Comment by jbrockmendel** [comment]
> If so you might want to opt for a macro'ed approach similar to what's in khash.h rather than copy / pasting these.

my thought was to follow the pattern for the dt64 code in the same file, which would probably mean refactoring out some helper functions but not using macros.  Is getting comfortable with macros the next step of leveling up my C-fu?

**Comment by WillAyd** [comment]
The macros would definitely be more advanced, at least as far as I understand. I am by no means an expert and not in a good enough position to teach, but from my personal experience the biggest things I've found to help grasp C are:

  - Treating every warning as an error. -Wall, -Wextra and -Werror are great switches; while C has some historic cruft in what the language will allow, most compilers do a great job of warning you about things that may be technically valid but not what you want
  - Getting comfortable with the debugger. gdb and lldb are great tools, and aren't a huge leap from pdb. The only thing to be aware of here is that if you want to debug a library you will also want to compile that library with debugging enabled, otherwise a lot of the function names, symbols and bodies will be stripped down to assembly

https://pandas.pydata.org/pandas-docs/stable/development/debugging_extensions.html might also help

**Comment by jbrockmendel** [comment]
Thanks.  Any objection to merging as-is?  We may implement more units later, but this gets all the ones we're definitely doing.

**Comment by jbrockmendel** [comment]
i guess it would make sense to have a corresponding function to go from tdstruct to td64. was that just not necessary at the time when it was nanos-only?

**Comment by WillAyd** [comment]
None from me. My feedback is just long term consideration but I think it’s great to move forward with what’s there and consolidate later

**Review by mroeschke — APPROVED** [review]
Not user facing necessarily right?

**Comment by jbrockmendel** [comment]
> Not user facing necessarily right?

nope, just needed to implement non-nano Timedelta

**Pull request merged**